### ISSUE TITLE
Enabling FIPS-compliant encryption

### ISSUE THREAD

**Issue by nzakas** [nodejs/node-v0.x-archive#25463]
For many in enterprises, regulations require the use of FIPS-compliant encryption tools. OpenSSL itself is not FIPS compliant, but does have a module that is (http://www.openssl.org/docs/fips/fipsnotes.html). It's possible to hack this into a custom build of Node.js (http://arthur-notes.youramaryllis.com/2014/12/nodejs-fips.html), but it would be nice to have it supported by Node directly (preferably starting with 0.10.x).

Someone had submitted a PR a while back for this: https://github.com/joyent/node/pull/6380

Is this something that would be considered?

**Comment by mhdawson** [comment]
I have had a few internal teams within IBM ask for this as well.  @nzakas have you looked into the details whether simply enabling the FIPs mode in OpenSSL would be enough for consumers to use Node when FIPS compliance is required ?  If you have researched/thought about this I'd be interested in setting up a time for a conversation.

**Comment by nzakas** [comment]
Are you talking about calling [`FIPS_mode_set(1)`](https://wiki.openssl.org/index.php/FIPS_mode_set%28%29) from somewhere?

**Comment by mhdawson** [comment]
No,  I meant does simply having Node use a version of OpenSSL compiled in FIPs mode enough for users that have a FIPs requirement as there is Node specific code in the path between the request and the crypto operations that take place.

**Comment by mhdawson** [comment]
The other issue might also be if there are any patches in Node which are floated on top of OpenSSL and what that would to do the FIPs cert

**Comment by indutny** [comment]
Hello!

I have did it in bud:

* https://github.com/indutny/bud/commit/bfece8c5c2c33e5d661b450e5395a89f3661c564
* https://github.com/indutny/bud/blob/master/fips.gypi

It should be pretty straightforward to integrate it in node.js/io.js .

The way it works in bud is following:

1. Build the fipscanister thing in some folder.
2. Supply this folder to configure: `./configure --fips-dir=...`
3. Build everything and enjoy!

There is only minor nit that will be required for this in node.js, that is not required in bud. It is patched version of `fipsld` that wraps premain.c with `-x c` and `-x none`, to make sure it works with C++ code.

**Comment by indutny** [comment]
TL;DR: going to port this code right now, expect a PR today or tomorrow.

**Comment by nzakas** [comment]
@mhdawson I believe so, I'll get clarification from our security folks. From my discussions with them thus far, it seems to be enough of we can prove Node.js crypto is running against the FIPS-enabled OpenSSL.

**Comment by indutny** [comment]
Yay, a PR for io.js: https://github.com/nodejs/io.js/pull/1890 . I guess it can be easily backported to node.js.

**Comment by nzakas** [comment]
Just confirmed with our security folks that we must call `FIPS_mode_set(1)` to be in compliance. I'm going to ask them to chime in directly.

**Comment by nc-box** [comment]
Going off of the [OpenSSL Fips User Guide](https://openssl.org/docs/fips/UserGuide-2.0.pdf), it should call ```FIPS_mode_set(1)``` in order to enable FIPS mode. This will begin all the tests and disable algorithms that are not FIPS validated. There's also ```FIPS_selftest_failed()``` which will return if the self tests succeeded or not. I couldn't confirm if this is required, but it's useful to call along with ```FIPS_mode_set```. 

Additionally as described above fipsld is required to properly do the linking and for the integrity tests.

Also I want to point out one important fact listed in page 42 and 45 of the user guide. Basically NIST standards maintain that the source must be obtained via a "secure installation". You cannot just download the source and validate it using a preinstalled FIPS validated openssl. Their recommendation is to contact OSF and ask for a CD with the source. Page 88 has alternative suggestions on how to validate it but even then no luck using an old OpenSSL version.

**Comment by jasnell** [comment]
@indutny :+1: 

@nzakas ... the key question for whether this can be enabled in v0.10 is whether it changes any of the default APIs or behaviors (which I wouldn't expect). The fact that it's put behind a `configure` flag makes it even better. It shouldn't be too difficult to backport it. @joyent/node-coreteam ... what do you think?

@nzakas .. interesting in helping out with a pull request based on @indutny's work? :-)

**Comment by mhdawson** [comment]
@nc-box  I think what we'd want is a documented set of steps that would result in a configuration that would meet FIPs requirements.  1) Get CD from X,  etc.

**Comment by nzakas** [comment]
@jasnell interested, yes, but I'm away from my computer for a few weeks (on a tablet right now). If someone else wants to take it, I'd be grateful; if not, I'll follow up when I'm back on a computer.

**Comment by admingermany** [comment]
Guten Morgen. . . . Kan mir einer verraten wo und wie ich ein aktewierungskey Finde ich such schon wie nen irren seid Tagen.

**Comment by nc-box** [comment]
@mhdawson At this time, I would suggest referring to the OpenSSL FIPS User Guide (https://openssl.org/docs/fips/UserGuide-2.0.pdf) for details around building a version of OpenSSL that meets the FIPS requirements. The User Guide provides more information and considerations when trying to build and use the FIPS compliant version of OpenSSL.

**Comment by arthurtsang** [comment]
so i was the author of PR #6380, anyway, this is how we build FIPS compliant nodejs http://arthur-notes.youramaryllis.com/2014/12/nodejs-fips.html.  and one "side" effect after enabling FIPS is calling anything function using md5 will crash.  node would just crash as FIPS implementation will exit the application.  btw, it's nice we finally have some discussion of getting FIPS compliance to nodejs.  thanks.

**Comment by indutny** [comment]
@arthurtsang nice one! I think we already landed it in io.js, see the commit above ;)

**Comment by ScarletTanager** [comment]
I'm one of the IBM internal users to whom @mhdawson referred above.  @nc-box is correct in that building a FIPS-enabled OpenSSL requires following very explicit instructions as detailed in the User Guide for the FIPS module.  It's not hard to do at all, and you can build either an archive or shared library (I've done the latter for use in building FIPS-capable Ruby, e.g.).  I have to admit that I (and a couple of security guys) missed the requirement to get the source (for both the FIPS module and OpenSSL, I assume) on CD, glad that got pointed out here.

Having said that, what is the status of this item?  As others have noted, we'd need a Node API call to invoke the FIPS_set_mode(1) library call.  I'm also curious about the question @mhdawson asked re: patches on top of OpenSSL (I apologize, but I'm not a Node.js guy, and I'm going based off of https://github.com/joyent/node/wiki/OpenSSL-upgrade-process).

**Comment by nzakas** [comment]
It's waiting for someone to port @indutny 's io.js patch to Node.

**Comment by indutny** [comment]
@nzakas why should it be ported? Technically it is the same project now, we're just merging everything together atm.

**Comment by jasnell** [comment]
Ok, so the question then becomes: is there reason to enable FIPS in v0.12. Technically that would qualify as a new feature. The informal policy so far has been to defer new features to the converged repo. Perhaps for v0.12, the best thing to do would be to document *how* someone could build v0.12 on their own with FIPS enabled but defer actually merging that into core until we're working from the converged repo (or, perhaps, getting it landed in the io.js stream).

**Comment by indutny** [comment]
@jasnell I believe it was already landed in io.js master some time ago.

**Comment by nzakas** [comment]
At work we are still using 0.10.x, upgrading breaks a lot of the world for us, so was looking for an intermediate solution.

**Comment by weagle08** [comment]
What is the status of this issue in the NodeJS 4.0 release? We are looking to use NodeJS on a project but it requires FIPS compliance.

**Comment by indutny** [comment]
It is already in 4.0

**Comment by weagle08** [comment]
sweet! thanks!

**Comment by arash16** [comment]
I'm using node-v5.0.0, after enabling fips support, as @arthurtsang said before any usage of md5 will crash, even the npm is broken..
I think it would be better to implement md5 inside crypto module on it's own.. There's a lot of frameworks depending on md5 for tag/crc generation (ex. Express.js)

![screenshot](https://cloud.githubusercontent.com/assets/4003500/10986085/245a917a-843c-11e5-9a06-2a55e51ee684.png)

**Comment by ScarletTanager** [comment]
@arash16 What do you mean, "enabling FIPS support?"  MD5 is not a FIPS-approved digest algorithm, so if you put OpenSSL into FIPS mode, MD5 is not going to be available to you.  In FIPS mode you need to use SHA.

**Comment by mhdawson** [comment]
I thinking at this point this issue can be closed.  Support is enabled in 4.x and later, and with the 0.10 and 0.12 end of life within 6 months I don't think it will be backported.  Please let me know if you disagree.
